test/widgets/store_test.dart

```diff
@@ -113,4 +150,65 @@ void main() {
     check(tester.any(find.textContaining('found store'))).isTrue();
     tester.widget(find.text('found store, account: ${eg.selfAccount.id}'));
   });
+
+  testWidgets('PerAccountStoreAwareStateMixin', (tester) async {
+    final widgetWithMixinKey = GlobalKey<MyWidgetWithMixinState>();
+    final accountId = eg.selfAccount.id;
+
+    await TestZulipBinding.instance.globalStore.add(eg.selfAccount, eg.initialSnapshot());
+
+    Future<void> pumpWithParams({required bool light, required int accountId}) async {
+      await tester.pumpWidget(
+        MaterialApp(
+          theme: light ? ThemeData.light() : ThemeData.dark(),
+          home: GlobalStoreWidget(
+            child: PerAccountStoreWidget(
+              accountId: accountId,
+              child: MyWidgetWithMixin(key: widgetWithMixinKey))),
+        ));
+    }
+
+    // [onNewStore] called initially
+    await pumpWithParams(light: true, accountId: accountId);
+    await tester.pump(); // global store
+    await tester.pump(); // per-account store
+    check(widgetWithMixinKey).currentState.isNotNull().storeChangeCounter.equals(1);
+    check(widgetWithMixinKey).currentState.isNotNull()
+      ..anyDepChangeCounter.equals(1)
+      ..storeChangeCounter.equals(1);
+
+    // [onNewStore] not called on unrelated dependency change
+    await pumpWithParams(light: false, accountId: accountId);
+    await tester.pumpAndSettle();
+    check(widgetWithMixinKey).currentState.isNotNull()
+      ..anyDepChangeCounter.equals(3) // (Dunno why 3 instead of 2. Anyway, more than 1…)
```

Comment:
I was curious too, so I tried setting `debugPrintRebuildDirtyWidgets = true;` before this `pumpAndSettle` call.

(Actually I started by figuring there'd be *some* debug variable I could turn on to get some relevant logging for this; so I typed "debug" and then browsed through the IDE's suggestions. The other one I tried was `debugPrintScheduleBuildForStacks`, and I think either one would have found the answer.)

(More accurately, I *started* by inserting `print(StackTrace.current);` into `MyWidgetWithMixinState.didChangeDependencies`. But the call stack here is generic framework machinery that doesn't really illuminate this question.)

The thing that kicks the dependency change off is:
```
Building AnimatedTheme(duration: 200ms, dirty, state: _AnimatedThemeState#0c4a7(ticker active, ThemeDataTween(ThemeData#187b1 → ThemeData#2deee)))
```

Note the "animated" in the name. Apparently by default a theme change happens over 200ms.

And then `pumpAndSettle` has a default per-frame duration of 100ms. So we get one frame where the theme is halfway between light and dark, and then one where it's dark.

Knowing that: to make this a bit cleaner, you can replace the `pumpAndSettle` with a single `pump(const Duration(seconds: 1))` (or any other duration of at least 200ms).

Reply:
```
Building AnimatedTheme(duration: 200ms, dirty, state: _AnimatedThemeState#0c4a7(ticker active, ThemeDataTween(ThemeData#187b1 → ThemeData#2deee)))
```

Interesting, thanks for the tip!

I'm curious; to me, that line is kind of a needle in a haystack, when I look at it with all the additional output I get:

[debugoutput.txt](https://github.com/zulip/zulip-flutter/files/12223201/debugoutput.txt)

Did it just stand out to you because of the "200ms"?

Reply:
Good question. The output I was looking at looked different, and I guess there are two things there:

One is that I only set those variables just before the `pumpAndSettle` in question, so all the setup up to that point wouldn't cause logging:
```diff
     // [onNewStore] not called on unrelated dependency change
     await pumpWithParams(light: false, accountId: accountId);
+    debugPrintRebuildDirtyWidgets = true;
     await tester.pumpAndSettle();
```

Then that line is the very first one, so it naturally stands out for that reason.

The other is that perhaps those other two kinds of logging I mentioned above were helpful after all, just for punctuating what's happening. In particular this:
```diff
@@ -28,6 +28,7 @@ class MyWidgetWithMixinState extends State<MyWidgetWithMixin> with PerAccountSto
 
   @override
   void didChangeDependencies() {
+    print(StackTrace.current);
     super.didChangeDependencies();
     anyDepChangeCounter++;
   }
```

So then those stack traces stand out, and they mark where the test widget is getting rebuilt. At the end of some of them, it looks like
```
[… bunch more stack frames …]
#20     TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1008:5)
<asynchronous suspension>
#21     StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:114:42)
<asynchronous suspension>

Building Text("brightness: Brightness.dark; accountId: 1001", dependencies: [DefaultSelectionStyle, DefaultTextStyle, MediaQuery])
Rebuilding AnimatedTheme(duration: 200ms, dirty, state: _AnimatedThemeState#1d770(ticker active, ThemeDataTween(ThemeData#b08bd → ThemeData#fb850)))
Rebuilding Theme(ThemeData#fb850, dependencies: [DefaultSelectionStyle])
[… bunch more rebuilds ….]
```

So the test widget is getting rebuilt, and then the `Text` that's its child… and then the next thing is this `AnimatedTheme`, so that must be what prompted a fresh round of rebuilds.


Here's the resulting full output (from `flutter test test/widgets/store_test.dart >& output.log`):
[output.log](https://github.com/zulip/zulip-flutter/files/12223971/output.log)